librz/bin/format/elf/elf_misc.c

```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: 2021 08A <08A@riseup.net>
+// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+// SPDX-License-Identifier: LPGL-3.0-only
+
+#include "elf.h"
+
+static bool buffer_read_8(ELFOBJ *bin, ut64 *offset, ut8 *result) {
+	ut8 tmp = rz_buf_read8_at(bin->b, *offset);
+	if (tmp == UT8_MAX) {
+		return false;
+	}
+
+	*offset += sizeof(ut8);
+	*result = tmp;
+
+	return true;
+}
+
+static bool buffer_read_16(ELFOBJ *bin, ut64 *offset, ut16 *result) {
+	ut16 tmp = rz_buf_read_ble16_at(bin->b, *offset, bin->endian);
+	if (tmp == UT16_MAX) {
+		return false;
+	}
+
+	*offset += sizeof(ut16);
+	*result = tmp;
+
+	return true;
+}
+
+static bool buffer_read_32(ELFOBJ *bin, ut64 *offset, ut32 *result) {
+	ut32 tmp = rz_buf_read_ble32_at(bin->b, *offset, bin->endian);
+	if (tmp == UT32_MAX) {
+		return false;
+	}
+
+	*offset += sizeof(ut32);
+	*result = tmp;
+
+	return true;
+}
+
+static bool buffer_read_32_signed(ELFOBJ *bin, ut64 *offset, st32 *result) {
+	ut32 tmp;
+
+	if (!buffer_read_32(bin, offset, &tmp)) {
+		return false;
+	}
+
+	*result = convert_to_two_complement_32(tmp);
+
+	return true;
+}
+
+static bool buffer_read_64(ELFOBJ *bin, ut64 *offset, ut64 *result) {
+	ut64 tmp = rz_buf_read_ble64_at(bin->b, *offset, bin->endian);
+	if (tmp == UT64_MAX) {
+		return false;
+	}
+
+	*offset += sizeof(ut64);
+	*result = tmp;
+
+	return true;
+}
+
+static bool buffer_read_64_signed(ELFOBJ *bin, ut64 *offset, st64 *result) {
+	ut64 tmp;
+
+	if (!buffer_read_64(bin, offset, &tmp)) {
+		return false;
+	}
+
+	*result = convert_to_two_complement_64(tmp);
```

Comment:
Mmmm I'm not convinced of this. Why is it necessary? Can't you directly read in a st64 variable?

Reply:
The conversion between signed and unsigned is platform dependent. So Rizin should manually convert with this function in such case.

Reply:
how is this done everywhere else? I think this is not a elf specific issue. @wargio @XVilka @thestr4ng3r

Reply:
I am not 100% sure, it's either not handled at all, or just some band-aid in particular places.

Reply:
Afaik in other cases it is just casted, which as @08A is implementation-defined (not undefined) behavior. So to be 100% correct, the approach here is indeed the right one.